service/tbc/peer/rawpeer/rawpeer_test.go

```diff
@@ -4,4 +4,133 @@
 
 package rawpeer
 
-// XXX add tests here
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/phayes/freeport"
+)
+
+func GetFreePort() string {
+	port, err := freeport.GetFreePort()
+	if err != nil {
+		panic(err)
+	}
+	return strconv.Itoa(port)
+}
+
+func mockPeerServer(ctx context.Context, id int, listener net.Listener, msgCh chan string) error {
+	conn, err := listener.Accept()
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	p, err := NewFromConn(conn, wire.TestNet3, wire.AddrV2Version, id)
+	if err != nil {
+		return err
+	}
+
+	var rmsg wire.Message
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			rmsg, _, err = p.Read(0)
+			if err != nil {
+				return err
+			}
+			msgCh <- rmsg.Command()
+		}
+	}
+}
+
+func TestConcurrentReadWrite(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	msgCh := make(chan string)
+
+	port := GetFreePort()
+
+	listener, err := net.Listen("tcp", ":"+port)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer listener.Close()
+
+	go func() {
+		err := mockPeerServer(ctx, 2, listener, msgCh)
+		if err != nil && !errors.Is(err, context.Canceled) && !errors.Is(err, io.EOF) {
+			panic(err)
+		}
+	}()
+
+	conn, err := net.Dial("tcp", "localhost:"+port)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer conn.Close()
+
+	p1, err := NewFromConn(conn, wire.TestNet3, wire.AddrV2Version, 1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := p1.Write(0, wire.NewMsgPing(1)); err != nil {
+		panic(err)
+	}
+
+	writerNum := 2
+	WriteCount := 100
```

Comment:
nit: I think these could be constants - Also, is there any reason `WriteCount` starts with a capitalised letter? 🤔 
```suggestion
	const (
		writeNum = 2
		writeCount = 100	
	)
```

Reply:
Yep makes sense. And no, just an accident 😄